geiger/src/graph/Graph.js

```diff
@@ -0,0 +1,226 @@
+/**
+ * Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Collection from '../util/Collection';
+import Events from '../util/Events';
+import Node from './Node';
+
+/**
+ * @typedef {import('../graph/Cell').default} Cell
+ * @typedef {import('../graph/Edge').default} Edge
+ */
+
+/**
+ * A module that tracks all the nodes and edges.
+ *
+ * @class Graph
+ * @extends Events
+ */
+export default class Graph extends Events {
+  constructor() {
+    super();
+
+    this._nodes = new Collection(null, {trackExit: true});
+    this._edges = new Collection(null, {trackExit: true});
+
+    // Cache the removed nodes/edges in current batch.
+    this._removed_nodes = [];
+    this._removed_edges = [];
+
+    /**
+     * For each node ID, keep track of all out-bound edge IDs.
+     * @type {!Object<string, !Object<string, true>>}
+     * @private
+     */
+    this._out = {};
+
+    /**
+     * For each node ID, keep track of all in-bound edge IDs.
+     * @type {!Object<string, !Object<string, true>>}
+     * @private
+     */
+    this._in = {};
+  }
+
+  /**
+   * Add cell to the graph based on its type.
+   * @param {!Cell} cell
+   */
+  addCell(cell) {
+    if (!cell.graph) {
+      cell.graph = this;
```

Comment:
This is not a good practice: Cell can add a graph into it, and also graph add a cell into itself. Let's make this a one-directional operation.

When you do `new Cell(graph)`, all of things should be triggered.

Reply:
Removed.